[IMP] website_sale: let users create and edit custom product ribbons

sdegueldre opened the pull request:
Previously, there was only one static "Sale" ribbon. Users might want to
have many types of ribbons, for different levels of sales, or to
indicate that a product is sold out. This commit lets the user create
custom ribbons and use them on prodcuts on the shop page.

task-1981170

qsm-odoo commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -186,14 +191,129 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
+        this.ribbonName = 'Ribbon text';
+        this.ribbonColor = '#ff9c00';
+        this._toggleEditingUI(true);
+    },
```
Not sure I am a fan of using the option system for stuff like that (even `saveRibbon` and `deleteRibbon` are questionable but kinda necessary for the `async` parts)... but I guess it is ok. No real argument to avoid it... In this case, I would have rendered the 2 buttons in JS and attached event handlers through the event properties. Like a normal widget.

sdegueldre replied:
Not sure what you mean here with "stuff like that"

qsm-odoo replied:
Toggling another part of the UI instead of editing the DOM like options are meant to, but I guess it is not really different from "opening a modal to edit something".

qsm-odoo commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -236,29 +356,48 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
-    _renderCustomWidgets: async function (uiFragment) {
-        const checkboxes = [];
+    async _renderCustomXML(uiFragment) {
+        const $select = $(uiFragment.querySelector('we-select'));
```
Search the select by class name (add a class in the xml)

qsm-odoo commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -236,29 +356,48 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
+        }).then(async ribbons => {
+            this.ribbons = Object.fromEntries(ribbons.map(r => [r.id, r]));
+            for (const ribbon of ribbons) {
+                $select.append($(`<we-button data-set-ribbon="${ribbon.id}">${ribbon.name}</we-button>`));
```
If you are using jQuery then `append(string)` will be enough, instead of `append($(string))`

qsm-odoo commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -236,29 +356,48 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
+                return this.ribbonName;
+            default:
+                return this._super(...arguments);
+        }
```
I am not a fan of functions which mean to return something but are doing it in the else clause... Just convention but I would return `_super` at the end of the function and not have a `default` clause.

qsm-odoo commented on addons/website_sale/static/src/scss/website_sale.scss:
```diff
@@ -134,25 +134,22 @@ $o-wsale-products-layout-grid-gutter-width: min($grid-gutter-width / 2, $o-wsale
-    .ribbon {
-        font: bold 15px Sans-Serif;
-        color: white;
+    .o_wribbon {
```
`o_wribbon`? Either `o_wsale_ribbon` or `o_ribbon` as a generic website style.

sdegueldre replied:
Do you think we might want it as a generic website style? I can maybe see it used for website snippet options in the future so maybe?

qsm-odoo replied:
In fact, already used by events (example on /event/design-fair-los-angeles-2020-03-28-2020-04-01-1/page/introduction-design-fair-los-angeles). Could be good to share the same style in this PR directly. So yes, generic class in website.scss.

sdegueldre replied:
In event, the ribbon is not in the same corner. Should I use the bootstrap system with one class for the common style and one for the specific style? (o_ribbon, o_ribbon_top_left, o_ribbon_top_right, ...)

qsm-odoo replied:
Yes. Or move the event ones at the top-left, your call :+1:

sdegueldre replied:
Actually it seems like we have ribbon implementations all over the place. Should I refactor all the frontend ones to use this new implementation and leave the backend ones (eg, Archived, Paid invoices) untouched? (those use a widget to change the font size to make it fit inside the ribbon)

qsm-odoo replied:
Yes. If you want, you can ping the framework team once the task is merged to tell them we have better ribbons if they want to get inspired :)

sdegueldre replied:
In the end, there was only actual ribbons in event, so I just overrode the transform properties there so it's in the top right (top left was hiding too much of the sponsor logos). There are a few things though that have ribbon in their name but none of the ribbon mechanics, or are in the backend, so I didn't touch those.

qsm-odoo commented on addons/website_sale/views/templates.xml:
```diff
@@ -156,6 +153,10 @@
                 <div class="o_wsale_product_btn"/>
             </div>
+            <t t-set="color" t-value="td_product['ribbon']['color'] or ''"/>
+            <t t-set="text_color" t-value="td_product['ribbon']['text_color'] or ''"/>
```
```suggestion
            <t t-set="text_color" t-value="td_product['ribbon']['text_color']"/>
            <t t-set="text_color" t-value="text_color and ('color: %s;' % text_color) or ''"/>
```
so that we do not generate `color: ;` ?

qsm-odoo commented on addons/website_sale/views/templates.xml:
```diff
@@ -156,6 +153,10 @@
+            <t t-set="color" t-value="td_product['ribbon']['color'] or ''"/>
+            <t t-set="text_color" t-value="td_product['ribbon']['text_color'] or ''"/>
+            <t t-set="colorClass" t-value="color if color.startswith('bg-') else None"/>
+            <span t-attf-class="o_wribbon #{colorClass}" t-attf-style="color: #{text_color}; #{not colorClass and 'background-color:' + color}" t-esc="td_product['ribbon']['name'] or ''"/>
```
As discussed above, not robust enough to check that it starts with `bg-`. I would simply have two fields instead of one in the model: one for classes, one for css color. That way no need of check here and if we later want the option to make a ribbon bold with the bold class, it is possible (so name that field generically, not bg_class)

qsm-odoo commented on addons/website_sale/views/templates.xml:
```diff
@@ -201,7 +202,7 @@
             <t t-set="additional_title">Shop</t>
             <div id="wrap" class="js_sale">
-                <div class="oe_structure" id="oe_structure_website_sale_products_1"/>
+                <div class="oe_structure oe_structure_small" id="oe_structure_website_sale_products_1"/>
```
I removed this from the specs, it is the task of RVE and it's more that just smaller, best to avoid useless conflicts.

sdegueldre replied:
Yeah sbu told me and it's in a separate commit, I just need to drop it.

sdegueldre commented on addons/website_sale/security/ir.model.access.csv:
```diff
@@ -6,7 +6,8 @@ access_product_category_pos_manager,product.public.category manager,model_produc
 access_product_pricelist_item_public,product.pricelist.item.public,product.model_product_pricelist_item,,1,0,0,0
-access_product_style,product.style.public,website_sale.model_product_style,,1,0,0,0
+access_product_ribbon_public,product.ribbon.public,website_sale.model_product_ribbon,,1,0,0,0
+access_product_ribbon_pos_manager,product.ribbon.pos_manager,website_sale.model_product_ribbon,sales_team.group_sale_manager,1,1,1,1
```
I reused the term that was used for people who have the rights to manage product categories higher in the file. Should I change it to "sale_manager" like the group name itself?

qsm-odoo replied:
I guess yes

sdegueldre commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -186,14 +191,129 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
+                    'product_tmpl_id': this.productTemplateID,
+                    'ribbon_id': widgetValue || false,
+                },
+            });
```
Not sure what you mean here.

qsm-odoo replied:
Add a ribbon on a product, hit discard, the ribbon is still saved. The ribbon was saved directly so not possible to discard.

sdegueldre replied:
Right! Previously, it would reload the page as well. There is a task in india for something similar: task-1960780

I'll think about it, see if it can be done in a more generic way that can be used by other options that change config on the server.

qsm-odoo replied:
Yeah I know I have to take a look at it again, but it is not similar as in their case it is not possible to not reload. The task changed to only make the reload more pretty (odoo modal instead of ugly browser one).

As yours is adding or removing a generic website item (a ribbon) (so not reloading a complex grid with internal t-field items), it is more possible :)

sdegueldre commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -177,6 +179,9 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
             .toggleClass('d-none', listLayoutEnabled);
+        // Ribbons may have been edited or deleted in another products' option, need to make sure they're up to date
+        // Maybe the ribbons could be stored on a "global" object to avoid making a RPC every time.
+        this._rerenderXML();
```
Any particular reason to do it that way?

qsm-odoo replied:
Not ignore the async part of `_rerenderXML`, while waiting for better: acknowledge the async part of onFocus, onBlur, on... etc.

sdegueldre commented:
@qsm-odoo Alright, I made the whole thing discardable, it's quite a bit of code though. Basically a rudimentary database cache with a transaction system for only one model. Was fun to implement though.

qsm-odoo commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -186,14 +365,82 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
+        const ribbon = this.ribbons[widgetValue] || {name: '', color: '', text_color: '', html_class: ''};
+        const $ribbons = $(`[data-ribbon-id="${widgetValue}"] .o_ribbon:not(.o_wsale_ribbon_dummy)`);
+        $ribbons.text(ribbon.name);
+        $ribbons.attr('class', 'o_ribbon');
```
Would this be that costly to support custom code adding custom classes here? Just remove all html_class of all ribbons?

qsm-odoo commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -186,14 +365,82 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
+        $ribbons.text(ribbon.name);
+        $ribbons.attr('class', 'o_ribbon');
+
+        if (!ribbon.html_class) {
+            $ribbons.css('background-color', ribbon.color);
+        } else {
+            $ribbons.addClass(ribbon.html_class);
+            $ribbons.css('background-color', '');
+        }
```
```javascript
$ribbons.css('background-color', ribbon.color ? ribbon.color : '');
if (ribbon.html_class) {
     $ribbons.addClass(ribbon.html_class);
}
```
would that be ok ? Would be more generic and allow both a class and a color as it may be the case in a future implementation I would like to do.

qsm-odoo commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -186,14 +365,82 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
+    async deleteRibbon(previewMode, widgetValue, params) {
+        const {ribbonId} = this.$target[0].dataset;
+        this.trigger_up('delete_ribbon', {id: ribbonId});
+        await this._rerenderXML();
```
Still don't like a method option being in charge of rerendering the option UI by itself. Can't you do that in updateUI like I said before?

qsm-odoo commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -186,14 +365,82 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
+    /**
+     * @see this.selectClass for params
+     */
+    setRibbonColor(previewMode, widgetValue, params) {
+        this.selectStyle.call(Object.assign({}, this, {$target: this.$ribbon}), previewMode, widgetValue, params);
+    },
```
Why is this not a selectStyle method with an applyTo?

qsm-odoo commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -186,14 +365,82 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
+     * @see this.selectClass for params
+     */
+    setRibbonName(previewMode, widgetValue, params) {
+        this.$ribbon.text(widgetValue);
```
Could be a `data-select-property="" data-property-name="textContent"`, less code

sdegueldre replied:
As discussed, since the commit that implements that isn't merged yet, and doesn't implement a computeWidgetState method which I need, I'll leave it as is for now.

qsm-odoo commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -227,6 +474,10 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
 
         // Adapt size array preview to fit ppr
         $size.find('tr td:nth-child(n + ' + parseInt(this.ppr + 1) + ')').hide();
+        if (this.rerender) {
+            this.rerender = false;
+            return this._rerenderXML();
+        }
     },
```
You actually did it in fact, why not `this.rerender = true` where needed then? ^^

sdegueldre replied:
I think I just got distracted in the middle of implementing that ^^'

qsm-odoo commented on addons/website_sale/views/templates.xml:
```diff
@@ -156,6 +153,10 @@
+            <t t-set="color" t-value="td_product['ribbon']['color'] or ''"/>
+            <t t-set="text_color" t-value="td_product['ribbon']['text_color']"/>
+            <t t-set="bg_class" t-value="td_product['ribbon']['html_class']"/>
+            <span t-attf-class="o_ribbon #{bg_class}" t-attf-style="#{text_color and ('color: %s; ' % text_color)}#{not bg_class and 'background-color:' + color}" t-esc="td_product['ribbon']['name'] or ''"/>
```
instead of `not bg_class and`, `color and`

sdegueldre commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -54,6 +54,177 @@ odoo.define('website_sale.editor', function (require) {
+            if (created.includes(ribbon)) {
+                return false;
+            }
+            for (const entry of Object.entries(ribbon)) {
+                const [key, value] = entry;
```
I wanted to do that initially and it didn't seem to work. Probably messed something else up, since this seems to be valid syntax.

sdegueldre commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -54,6 +54,177 @@ odoo.define('website_sale.editor', function (require) {
+            if (created.includes(ribbon)) {
+                return false;
+            }
+            for (const entry of Object.entries(ribbon)) {
+                const [key, value] = entry;
+                if (value !== (this.originalRibbons[ribbon.id][key])) {
+                    return true;
+                }
+            }
+            return false;
```
I just did not know some was standard. That's much better.

sdegueldre commented on addons/website_event_track/static/src/css/website_event_track.css:
```diff
@@ -1,32 +1,25 @@
   -moz-box-shadow: 0px 0px 3px rgba(0, 0, 0, 0.3);
   box-shadow: 0px 0px 3px rgba(0, 0, 0, 0.3);
+
+  /* Move ribbon to top right corner */
+  right: 0;
+  left: unset;
+  transform-origin: top right;
+  transform: translateX(calc(-0.708 * (100% - 200px))) rotate(45deg) translateX(calc(100% - 100px));
```
I'd rather use two classes per ribbon (o_ribbon + o_ribbon_left/o_ribbon_right) so that I can use the common class to determine if something has a ribbon at all, if you don't mind. And apparently, there are ribbons on both the left and the right in @Cocographique 's prototype for the new shop page so it may not be a bad thing to have both available in website.

qsm-odoo replied:
Let's go with that then yes

sdegueldre commented on addons/website_sale/models/product.py:
```diff
@@ -187,7 +189,7 @@ class ProductTemplate(models.Model):
     website_size_x = fields.Integer('Size X', default=1)
     website_size_y = fields.Integer('Size Y', default=1)
-    website_style_ids = fields.Many2many('product.style', string='Styles')
+    website_ribbon_id = fields.Many2one('product.ribbon', string='Ribbon')
```
Sure, but what should the script even do? Just remove and add the required tables and columns, or something more?

qsm-odoo replied:
https://github.com/odoo/upgrade/wiki/How-To#when-to-write-upgrade-scripts

New tables are not listed, so only removing this field and adding the new one. You should also use the "promo" default ribbon for the product using the old ribbon style, that way people do not lose their ribbon at migration.

There is also "removed records" in the list, so I guess removing all product style record applies.

sdegueldre commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -54,6 +54,177 @@ odoo.define('website_sale.editor', function (require) {
+                fields: ['id', 'name', 'color', 'text_color', 'html_class'],
+            });
+        }
+        this.ribbons = Object.fromEntries(ribbons.map(ribbon => [ribbon.id, ribbon]));
```
Right, I'll try to keep that in mind :)

sdegueldre commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -170,13 +341,21 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
+    /**
+     * @override
+     */
+    start: function () {
+        this.$ribbon = this.$('.o_ribbon').clone().addClass('d-none o_wsale_ribbon_dummy').appendTo(this.$target);
+    },
```
Nothing to quip about here, I just forgot :+1:

sdegueldre commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -54,6 +54,177 @@ odoo.define('website_sale.editor', function (require) {
+        }, {});
+        // Inverting the relationship so that we have all templates that have the same ribbon to reduce RPCs
+        const ribbonTemplates = Object.entries(finalTemplateRibbons).reduce((acc, [templateId, ribbonId]) => {
+            acc[ribbonId] = [parseInt(templateId)].concat(acc[ribbonId] || []);
```
Have had to do this kind of thing a lot in coding golf challenges, so I just got used to this form I guess, but I have to agree it's way less readable, and the concat makes it slow for large inputs. Changed it to the second form :)

sdegueldre commented:
@qsm-odoo everything done except the migration script, will be working on that next

qsm-odoo commented on addons/website_sale/views/templates.xml:
```diff
@@ -234,7 +235,8 @@
                                                     <td t-att-colspan="td_product['x'] != 1 and td_product['x']"
                                                         t-att-rowspan="td_product['y'] != 1 and td_product['y']"
-                                                        t-attf-class="oe_product #{td_product['class']}">
+                                                        class="oe_product"
```
After discussion with JKE, as this has high chances of user customization + we customize it ourself and probably again in the future refactoring of shop, let's leave t-attf-class even if nothing is computed. Maybe just add a comment alongside.

qsm-odoo commented on addons/website_sale/views/snippets.xml:
```diff
@@ -62,9 +62,21 @@
+            </we-row>
+            <we-row class="d-none" data-name="ribbon_customize_opt">
+                <we-input string="Ribbon" data-set-ribbon-name="" data-apply-to=".o_wsale_ribbon_dummy" data-no-preview="true"/>
+                <we-colorpicker data-select-style="" data-apply-to=".o_wsale_ribbon_dummy" data-css-property="color" data-no-preview="true"/>
+                <we-colorpicker data-select-style="" data-apply-to=".o_wsale_ribbon_dummy" data-css-property="background-color" data-color-prefix="bg-" data-no-preview="true"/>
```
Add a `title=""` on those two colorpicker and I would invert their order (the recommended use case being choose a background color class and not choosing a text color).

sdegueldre commented:
@qsm-odoo made all adaptations and fixed the bug you reported, added the debug-mode many2one for form view in a separate commit for now while waiting for JKE's opinion.

JKE-be commented:
yes ok for me to keep it in backend ;)

sdegueldre commented:
@qsm-odoo added the requested features, cleaned some things up (renamed some fields to make more sense since we're now allowing HTML, etc) and rebased on master.

qsm-odoo commented on addons/website_sale/static/src/js/website_sale.editor.js:
```diff
@@ -236,28 +537,75 @@ options.registry.WebsiteSaleProductsItem = options.Class.extend({
+                .split(' ')
+                .filter(className => !/^o_(ribbon|tag)_(left|right)$/.test(className))
+                .join(' ');
+            $select.append(`<we-button data-set-ribbon="${ribbon.id}">
+                ${ribbon.html}
+                <span class="fa fa-${/o_tag_(left|right)/.test(ribbon.html_class) ? 'tag' : 'bookmark'} ml-auto"></span>
+                <span class="fa fa-arrow-${/o_(tag|ribbon)_left/.test(ribbon.html_class) ? 'left' : 'right'} ml-1"></span>
+                <span class="o_wsale_color_preview ${colorClasses} ml-1" style="background-color: ${ribbon.bg_color}"></span>
+                <span class="o_wsale_color_preview ${colorClasses} ml-1" style="background-color: ${ribbon.text_color || colorClasses ? 'currentColor' : defaultTextColor} !important;"></span>
+            </we-button>`);
```
Seems enough xml to justify using qweb

qsm-odoo commented on addons/website_sale/views/product_views.xml:
```diff
@@ -94,7 +94,7 @@
                             <field name="alternative_product_ids" widget="many2many_tags" domain="[('id', '!=', active_id), '|', ('company_id', '=', company_id), ('company_id', '=', False)]"/>
                             <field name="accessory_product_ids" widget="many2many_tags"/>
-                            <field name="website_style_ids" widget="many2many_tags" groups="base.group_no_one"/>
+                            <field name="website_ribbon_id" widget="many2one" groups="base.group_no_one"/>
```
`widget="many2one"` is useless

qsm-odoo commented on addons/website_sale/views/snippets.xml:
```diff
@@ -62,8 +62,30 @@
+                    <we-button class="fa fa-check" data-save-ribbon="" title="Validate" data-no-preview="true"/>
+                    <we-button class="fa fa-trash" data-delete-ribbon="" title="Delete" data-no-preview="true"/>
+                </we-row>
+                <we-colorpicker string="Text Color" title="" data-select-style="" data-apply-to=".o_wsale_ribbon_dummy" data-css-property="color" data-no-preview="true"/>
+                <we-colorpicker string="BG Color" title="" data-select-style="" data-apply-to=".o_wsale_ribbon_dummy" data-css-property="background-color" data-color-prefix="bg-" data-no-preview="true"/>
```
- Switch bg and text colorpickers like I explained ;) Top widgets influence bottom widgets, not the opposite when possible.
- Rename "BG Color" into "Background Color"

qsm-odoo commented on addons/website_sale/views/templates.xml:
```diff
@@ -275,7 +278,7 @@
     <template id="products_images_full" name="Images Full" inherit_id="website_sale.products" active="False" customize_show="True">
         <xpath expr="//t[@t-foreach='tr_product']//td" position="attributes">
-            <attribute name="t-attf-class" add="oe_image_full" separator=" "/>
+            <attribute name="class" add="oe_image_full" separator=" "/>
```
Not working anymore, t-attf-class

sdegueldre commented:
@qsm-odoo fixed the requested points and rebased this and the theme branch on master, upgrade branch was rebased this morning so it should be good already.
